src/pip/_internal/req/req_install.py

```diff
@@ -118,17 +117,22 @@ def __init__(
         self.req = req
         self.comes_from = comes_from
         self.constraint = constraint
-        if source_dir is None:
-            self.source_dir = None  # type: Optional[str]
-        else:
-            self.source_dir = os.path.normpath(os.path.abspath(source_dir))
         self.editable = editable
 
         if link is None and req and req.url:
             # PEP 508 URL requirement
             link = Link(req.url)
         self.link = self.original_link = link
         self.original_link_is_in_wheel_cache = False
+
+        self.source_dir = None  # type: Optional[str]
+        if self.editable:
+            assert link
```

Comment:
For consistency with the conditional in the next block.

```suggestion
            assert self.link
```